Dockerfile

```diff
@@ -1,20 +1,32 @@
-FROM crystallang/crystal:1.2.2-alpine AS builder
+FROM docker.io/node:18-alpine AS node-builder
+RUN mkdir /build
+WORKDIR /build
+COPY --chown=0:0 package.json package-lock.json /build/
+RUN npm ci
+COPY --chown=0:0 webpack.config.js ./
+RUN mkdir src
+COPY --chown=0:0 src/assets src/assets
+RUN npm run build
```

Comment:
You can view this as one build environment dedicated for the node build part.

On line 4-5, we copy package.json and package-lock.json and then prepare the node_modules. Because only those two files have been copied, the layer will only be re-generated whenever those two file change, saving on the incremental compilation.

We chown the files to root, so they don't inherit the UID of the host system, leading to more reproducible builds.

Then we do the same and only copy the config and assets needed for the frontend build.